### PR TITLE
Add MemoryMappedBuffer.

### PR DESCRIPTION
**Pull request by hansenyy**
This PR adds an implementation of jsi::Buffer that is based on memory mapped files. It will be used to implement a ScriptStore that is based on memory mapped files.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/3059)

### CODE REVIEW

File: vnext/Desktop.DLL/react-native-win32.x64.def [hunk 1]
```diff
@@ -46,6 +46,7 @@ EXPORTS
 ?typeName@dynamic@folly@@QEBAPEBDXZ
 ?updateProperties@ShadowNode@react@facebook@@UEAAX$$QEBUdynamic@folly@@@Z
 ?utf16ToUtf8@unicode@react@facebook@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@AEBV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@5@@Z
+?makeMemoryMappedBuffer@jsi@Microsoft@@YA?AV?$shared_ptr@VBuffer@jsi@facebook@@@std@@PEB_WI@Z
```
Comment (stecrain) [inline vnext/Desktop.DLL/react-native-win32.x64.def]: is this export just for unit tests? do you want to add it for the uwp binary as well?

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  }
+
+  if (size.LowPart == 0) {
+    throw facebook::jsi::JSINativeException("Cannot memory map an empty file.");
+  }
+
+  if (size.HighPart != 0) {
+    throw facebook::jsi::JSINativeException(
+        "MemoryMappedBuffer only support files whose size can fit within an "
+        "uint32_t.");
```
Comment (stecrain) [inline vnext/Shared/MemoryMappedBuffer.cpp]: does this effectively limit to 4GB files? seems fine just curious if i'm understanding correctly :)

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+MemoryMappedBuffer::MemoryMappedBuffer(
+    const wchar_t *const filename,
+    uint32_t offset)
+    : m_fileMapping{nullptr, &CloseHandle},
+      m_fileData{nullptr, &fileDataDeleter},
+      m_size{0},
+      m_offset{0} {
+#if (defined(WINRT))
+  std::unique_ptr<void, decltype(&CloseHandle)> fileHandle{
+      CreateFile2(
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >CreateFile2 [](start = 6, length = 11)

Please add comment why we are using different Apis on Win32 and WinRt. Ideally, we would use the same Api

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+#include "MemoryMappedBuffer.h"
+
+#include <werapi.h>
+#include <windows.h>
+
+namespace {
+
+class MemoryMappedBuffer : public facebook::jsi::Buffer {
+ public:
+  MemoryMappedBuffer(const wchar_t *const filename, uint32_t offset);
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >const wchar_t *const [](start = 21, length = 20)

(nit) why not just const whar_t*?

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+  size_t size() const override;
+  const uint8_t *data() const override;
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >const uint8_t * [](start = 2, length = 15)

usually we would specify types as "const uint8_t* data()". Star and Apostrophe is right after type. Is fb convention required different syntax?

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+ private:
+  static void fileDataDeleter(void *p) {
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >void *p [](start = 30, length = 7)

(nit) "void* p" ?

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+const uint8_t *MemoryMappedBuffer::data() const {
+  return static_cast<const uint8_t *>(m_fileData.get()) + m_offset;
+}
+
+} // anonymous namespace
+
+namespace Microsoft::JSI {
+
+std::shared_ptr<facebook::jsi::Buffer> makeMemoryMappedBuffer(
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >makeMemoryMappedBuffer [](start = 39, length = 22)

(nit) MakeMemoryMappedBuffer

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  uint32_t m_size;
+  uint32_t m_offset;
+};
+
+MemoryMappedBuffer::MemoryMappedBuffer(
+    const wchar_t *const filename,
+    uint32_t offset)
+    : m_fileMapping{nullptr, &CloseHandle},
+      m_fileData{nullptr, &fileDataDeleter},
+      m_size{0},
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >{0} [](start = 12, length = 3)

Please make this assignment on line 24 during the declaration

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 6 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: The "void *p" style is enforced through clang format.

---
In reply to: [320883993](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320883993) [](ancestors = 320883993)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 5 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Yes, Facebook's convention is to put the asterisk next to the variable name, and this convention is enforced through clang format.

---
In reply to: [320883747](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320883747) [](ancestors = 320883747)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 4 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: The extra const makes sure that filename cannot be modified within the constructor function. It's just a slightly stronger const guarantee and I personally prefer to use it unless there is a reason not to.

---
In reply to: [320882711](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320882711) [](ancestors = 320882711)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 7 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Facebook's convention is to use camelCase instead of PascalCase for function names, and that is the convention I followed.

---
In reply to: [320884862](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320884862) [](ancestors = 320884862)

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+      &CloseHandle};
+#endif
+
+  if (fileHandle.get() == INVALID_HANDLE_VALUE) {
+    throw facebook::jsi::JSINativeException(
+        "CreateFileW/CreateFile2 failed with last error " +
+        std::to_string(GetLastError()));
+  }
+
+  LARGE_INTEGER size;
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >size [](start = 16, length = 4)

(nit) rename to fileSize

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        std::to_string(GetLastError()));
+  }
+
+  if (size.LowPart == 0) {
+    throw facebook::jsi::JSINativeException("Cannot memory map an empty file.");
+  }
+
+  if (size.HighPart != 0) {
+    throw facebook::jsi::JSINativeException(
+        "MemoryMappedBuffer only support files whose size can fit within an "
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >support [](start = 33, length = 7)

(nit) supports

File: vnext/Desktop.DLL/react-native-win32.x64.def (hunk 1 above)
Comment (hansenyy) [reply vnext/Desktop.DLL/react-native-win32.x64.def]: That is my mistake. I'll add them to the uwp binary too.

---
In reply to: [320814285](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320814285) [](ancestors = 320814285)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 8 above)
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >m_size [](start = 6, length = 6)

(nit) m_fileSize

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+      fileHandle.get(),
+      nullptr /* lpAttributes */,
+      PAGE_READONLY,
+      0 /* dwMaximumSizeHigh */,
+      m_size,
+      nullptr /* lpName */));
+#endif
+
+  if (!m_fileMapping) {
+    m_size = m_offset = 0;
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >m_size = m_offset = 0; [](start = 4, length = 22)

Why is this needed? We should just throw exception and leave object's state as is for debugging purposes.

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 8 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: I am more than happy to make this change. Out of curiosity though, what's the advantage for the assignment syntax?

---
In reply to: [320902973](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320902973) [](ancestors = 320902973)

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+  m_fileData.reset(MapViewOfFile(
+      m_fileMapping.get(),
+      FILE_MAP_READ,
+      0 /* dwFileOffsetHigh */,
+      0 /* dwFileOffsetLow */,
+      0 /* dwNumberOfBytesToMap */));
+#endif
+
+  if (!m_fileData) {
+    m_size = m_offset = 0;
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >m_size = m_offset = 0; [](start = 4, length = 22)

remove

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    throw facebook::jsi::JSINativeException("Invalid offset.");
+  }
+  m_offset = offset;
+
+#if (defined(WINRT))
+  m_fileMapping.reset(CreateFileMappingFromApp(
+      fileHandle.get(),
+      nullptr /* SecurityAttributes */,
+      PAGE_READONLY,
+      m_size,
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >m_size [](start = 6, length = 6)

From code, this seems to always equal to file size. If so, please rename it to m_fileSize

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    m_size = m_offset = 0;
+    throw facebook::jsi::JSINativeException(
+        "MapViewOfFile/MapViewOfFileFromApp failed with last error " +
+        std::to_string(GetLastError()));
+  }
+
+  WerRegisterMemoryBlock(m_fileData.get(), m_size);
+}
+
+size_t MemoryMappedBuffer::size() const {
```
Comment (NikoAri) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >size_t [](start = 0, length = 6)

it would be nicer to keep consistent types, either size_t or uint32_t

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+using facebook::jsi::Buffer;
+using facebook::jsi::JSINativeException;
+using facebook::react::utilities::checkedReinterpretCast;
+using Microsoft::JSI::makeMemoryMappedBuffer;
+using Microsoft::VisualStudio::CppUnitTestFramework::Assert;
+
+namespace {
+
+static constexpr const wchar_t *const testFileName = L"temp.txt";
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >testFileName [](start = 38, length = 12)

this should really be writing to temp directory instead of directory where process is executing

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+static void writeTestFile(const char *const content, size_t size) {
+  FILE *testFilePtr;
+
+  // We need pass in the "b" flag to ensure that fwrite modify its input when
+  // outputting to disk.
+  if (_wfopen_s(&testFilePtr, testFileName, L"wb")) {
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >_wfopen_s [](start = 6, length = 9)

this should crash, or at least throw exception

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    return;
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >return [](start = 4, length = 6)

crash or throw exception

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  std::unique_ptr<FILE, decltype(&fclose)> testFilePtrWrapper(
+      testFilePtr, fclose);
+
+  fwrite(content, sizeof(*content), size, testFilePtrWrapper.get());
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >fwrite [](start = 2, length = 6)

check for error and crash

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+static void deleteTestFile() {
+  if (!DeleteFileW(testFileName)) {
+    Assert::IsTrue(GetLastError() == ERROR_FILE_NOT_FOUND);
+  }
+  Assert::IsTrue(!PathFileExistsW(testFileName));
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >Assert::IsTrue [](start = 2, length = 14)

Asserts are used for verifying functionality of our code. Here, you are actually setting up a test, you are not writing a unit test for PathFileExists Api. Therefore, this should be crash or throw exception, instead of Assert.

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+   TEST_METHOD_INITIALIZE(initialize) {
+     deleteTestFile();
+   }
+
+   TEST_METHOD_CLEANUP(cleanUp) {
+     deleteTestFile();
+   }
+
+ public :
+  TEST_METHOD(SimpleTest_NoOffest) {
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >SimpleTest_NoOffest [](start = 14, length = 19)

Please be consistent with naming, some are Pascal (SimpleTest_NoOffset), some are Camel (cleanUp)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+    constexpr const char *const content =
+      "This is a very interesting string.";
+    const size_t size = strlen(content);
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >size [](start = 17, length = 4)

(nit) contentSize

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 15 above)
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >t t [](start = 36, length = 3)

TestFileName or s_testFileName or g_testFileName

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+    writeTestFile(content, size);
+
+    std::shared_ptr<Buffer> buffer = makeMemoryMappedBuffer(testFileName);
+
+    Assert::IsTrue(buffer->size() == size);
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >Assert [](start = 4, length = 6)

This Assert usage is correct!!!

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  TEST_METHOD(ErrorTest_InvalidOffset) {
+    writeTestFile("a", 1);
+
+    try {
+      size_t badOffset = 3;
+      std::shared_ptr<Buffer> buffer =
+        makeMemoryMappedBuffer(testFileName, badOffset);
+    } catch (JSINativeException& e) {
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >JSINativeException [](start = 13, length = 18)

You will also need to add verification that exception was actually thrown. Same comment for line 131.

File: vnext/Desktop.DLL/react-native-win32.x86.def [hunk 24]
```diff
@@ -46,6 +46,7 @@ EXPORTS
 ?typeName@dynamic@folly@@QBEPBDXZ
 ?updateProperties@ShadowNode@react@facebook@@UAEX$$QBUdynamic@folly@@@Z
 ?utf16ToUtf8@unicode@react@facebook@@YG?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@ABV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@5@@Z
+?makeMemoryMappedBuffer@JSI@Microsoft@@YG?AV?$shared_ptr@VBuffer@jsi@facebook@@@std@@QB_WI@Z
```
Comment (NikoAri) [inline vnext/Desktop.DLL/react-native-win32.x86.def]: >?ma [](start = 0, length = 3)

(nit) alphabetize

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 2 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Yes you are correct, and the old ChakraExecutor stack suffers from the same limitation.

---
In reply to: [320836911](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320836911) [](ancestors = 320836911)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 7 above)
Comment (JunielKatarn) [reply vnext/Shared/MemoryMappedBuffer.cpp]: In my experience, we've been using PascalCase for function names.
As far as I understand, we don't need to follow Facebook's conventions for the new code we are developing, but only when modifying their code bases (i.e. `react-native`).

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 14 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: This method is inherited from facebook::jsi::Buffer, so its return type has to be size_t. I tried to use uint32_t everywhere else because uint32_t's size is guaranteed.

---
In reply to: [320912391](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320912391) [](ancestors = 320912391)

File: vnext/ReactWindowsCore/MemoryMappedBuffer.h [hunk 25]
```diff
@@ -0,0 +1,15 @@
...
+
+#include "jsi/jsi.h"
+
+#include <memory>
+
+namespace Microsoft::JSI {
+
+// We only support files whose size can fit within an uint32_t. Memory
+// mapping an empty or a larger file fails.
+std::shared_ptr<facebook::jsi::Buffer> makeMemoryMappedBuffer(
```
Comment (JunielKatarn) [inline vnext/ReactWindowsCore/MemoryMappedBuffer.h]: Pascal convention, please.
`MakeMemoryMappedBuffer`.

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 3 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Added some comments; please review.

---
In reply to: [320882173](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320882173) [](ancestors = 320882173)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 8 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: I added default intializers in the class declaration.

---
In reply to: [320909429](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320909429) [](ancestors = 320909429,320902973)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 9 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Fixed.

---
In reply to: [320906933](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320906933) [](ancestors = 320906933)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 10 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Fixed.

---
In reply to: [320907580](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320907580) [](ancestors = 320907580)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 11 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: That makes sense. Fixed.

---
In reply to: [320909288](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320909288) [](ancestors = 320909288)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 12 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Fixed.

---
In reply to: [320909487](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320909487) [](ancestors = 320909487)

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 13 above)
Comment (hansenyy) [reply vnext/Shared/MemoryMappedBuffer.cpp]: Fixed.

---
In reply to: [320910059](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320910059) [](ancestors = 320910059)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 17 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: Fixed.

---
In reply to: [320914185](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320914185) [](ancestors = 320914185)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 18 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: Fixed.

---
In reply to: [320915218](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320915218) [](ancestors = 320915218)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 15 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: Changed to using a temp file in the temp directory.

---
In reply to: [320913312](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320913312) [](ancestors = 320913312)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 19 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: I changed it so that it crashes here instead.

---
In reply to: [320916475](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320916475) [](ancestors = 320916475)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 20 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: I went ahead ended up using PascalCase for everything.

---
In reply to: [320917305](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320917305) [](ancestors = 320917305)

File: vnext/ReactWindowsCore/MemoryMappedBuffer.h (hunk 25 above)
Comment (hansenyy) [reply vnext/ReactWindowsCore/MemoryMappedBuffer.h]: Fixed.

---
In reply to: [320977423](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320977423) [](ancestors = 320977423)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 15 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: Fixed.

---
In reply to: [320918079](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320918079) [](ancestors = 320918079)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 23 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: I changed the code to use Assert::ExpectException instead. Unfortunately, this also means that we can't check that the .what() string is what we expect.

---
In reply to: [320919530](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320919530) [](ancestors = 320919530)

File: vnext/Desktop.DLL/react-native-win32.x86.def (hunk 24 above)
Comment (hansenyy) [reply vnext/Desktop.DLL/react-native-win32.x86.def]: It looks like we need to do so for all the def files. I'll send out a separate PR for this.

---
In reply to: [320920138](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320920138) [](ancestors = 320920138)

File: vnext/Shared/MemoryMappedBuffer.cpp [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+const uint8_t *MemoryMappedBuffer::data() const {
+  return static_cast<const uint8_t *>(m_fileData.get()) + m_offset;
+}
+
+} // anonymous namespace
+
+namespace Microsoft::JSI {
+
+std::shared_ptr<facebook::jsi::Buffer> MakeMemoryMappedBuffer(
+    const wchar_t *const filename,
```
Comment (Khalef1) [inline vnext/Shared/MemoryMappedBuffer.cpp]: >filename [](start = 25, length = 8)

null check?

File: vnext/Shared/MemoryMappedBuffer.cpp (hunk 8 above)
Comment (NikoAri) [reply vnext/Shared/MemoryMappedBuffer.cpp]: The advantage is that it's easier to track where the default values are coming from. In this case, it doesn't matter too much, but imagine when there are multiple constructors, it could be difficult to ensure that object is initialized to same values in all of them.

---
In reply to: [320991285](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r320991285) [](ancestors = 320991285,320909429,320902973)

File: vnext/ReactUWP/EndPoints/dll/react-native-uwp.arm.def [hunk 27]
```diff
@@ -38,6 +38,7 @@ EXPORTS
 ?utf16ToUtf8@unicode@react@facebook@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@ABV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@5@@Z
 ?utf16ToUtf8@unicode@react@facebook@@YA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@ABV?$basic_string_view@_WU?$char_traits@_W@std@@@5@@Z
 ?utf8ToUtf16@unicode@react@facebook@@YA?AV?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@ABV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@5@@Z
+?MakeMemoryMappedBuffer@JSI@Microsoft@@YA?AV?$shared_ptr@VBuffer@jsi@facebook@@@std@@QB_WI@Z
```
Comment (NikoAri) [inline vnext/ReactUWP/EndPoints/dll/react-native-uwp.arm.def]: >?Ma [](start = 0, length = 3)

please alphabetize

Comment (NikoAri) [reply vnext/ReactUWP/EndPoints/dll/react-native-uwp.arm.def]: same for other def files

---
In reply to: [321853272](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r321853272) [](ancestors = 321853272)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 28]
```diff
@@ -0,0 +1,164 @@
...
+  TEST_METHOD(ErrorTest_InvalidOffset) {
+    WriteTestFile("a", 1);
+
+    Assert::ExpectException<JSINativeException>([this] {
+      uint32_t badOffset = 3;
+      std::shared_ptr<Buffer> buffer =
+          MakeMemoryMappedBuffer(m_testFileName.c_str(), badOffset);
+    });
+  }
+};
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: (nit) would be nice to place it into namespace per good C++ coding practice

File: vnext/ReactUWP/EndPoints/dll/react-native-uwp.arm.def (hunk 27 above)
Comment (hansenyy) [reply vnext/ReactUWP/EndPoints/dll/react-native-uwp.arm.def]: I would like send a separate PR for alphabetizing all the def files.

---
In reply to: [321853374](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r321853374) [](ancestors = 321853374,321853272)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp (hunk 28 above)
Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: It looks like none of our unit test code is in a namespace. I'll send a separate PR and make said change to all of them.

---
In reply to: [321853819](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r321853819) [](ancestors = 321853819)

File: vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp [hunk 29]
```diff
@@ -0,0 +1,164 @@
...
+ private:
+  std::wstring m_testFileName;
+
+  void WriteTestFile(const char *const content, size_t size) {
+    FILE *testFilePtr;
+
+    // We need pass in the "b" flag to ensure that fwrite modify its input when
+    // outputting to disk.
+    if (_wfopen_s(&testFilePtr, m_testFileName.c_str(), L"wb")) {
+      std::terminate();
```
Comment (NikoAri) [inline vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: >std::terminate(); [](start = 6, length = 17)

maybe this is not related to this particular CR. The problem is that std::terminate doesn't produce Watson dump containing callstack AFAIK. So investigating such failures will be very difficult, especially if they are intermittent and fail on Continuous Integration machines.

Either VEC or throwing C++ exception produces usable dump. You can leave this as is for this test, but we should address this at some point.

Comment (hansenyy) [reply vnext/Desktop.UnitTests/MemoryMappedBufferTests.cpp]: I agree this warrants a discussion.

---
In reply to: [321859108](https://github.com/microsoft/react-native-windows/pull/3059#discussion_r321859108) [](ancestors = 321859108)
